CliFx/Suggestions/Design.md

```diff
@@ -0,0 +1,272 @@
+
+Suggestions are given for only one type at a time; ie. commands and parameters are not suggested together. Suggestions are provided as follows:
+
+* Provides suggestions that **start with** the terms provided.
+* Don't provide other completions if there's already a match.
+* Be case insensitive when matching provided terms.
+
+
+# User Experience
+Auto-completions must be enabled by user as follows:
+
+1. Ensure that the executable is on the path
+2. Modify the terminal's user profile, adding a hook that calls CliFx for suggestions.
+3. Reload terminal profile.
+
+Users may use CliFx to install the suggestion hook automatically (step 2), or they do all three steps manually. Once installed, the user experience is varies depending on the terminal in use:
+
+### Powershell (windows):
+* tab to cycle through auto-completions
+* control-space to select from auto-completions
+```ps
+# when: user presses tab for auto-complete
+PS C:\> cli.exe <tab>
+# then: auto-complete fills in the very first match
+PS C:\> cli.exe book
+# when: user continues to press tab...
+PS C:\> cli.exe book<tab>
+# then: auto-complete cycles through each match
+PS C:\> cli.exe book add
+PS C:\> cli.exe book list
+PS C:\> cli.exe book remove
+
+# when: the user presses control-space for auto-complete
+PS C:\> cli.exe <ctrl-space>
+# then: the user is presented with a series of options to select from:
+PS C:\> cli.exe book
+book        book add        book list       book remove
+```
+
+### Powershell (ubuntu):
+* tab to view first auto-completion
+* tab twice to see possible auto-completions
+```ps
+# when: user presses tab for auto-complete
+PS C:\> cli.exe <tab>
+# then: auto-complete fills in the very first match
+PS C:\> cli.exe book
+# when: user continues to press tab...
+PS C:\> cli.exe book<tab>
+# then: the user is presented with all options
+book        book add        book list       book remove
+PS C:\> cli.exe book
+```
+
+### Bash (ubuntu):
+* tab to view first auto-completion
+* tab twice to see possible auto-completions
+```bash
+# when: user presses tab for auto-complete
+mauricel@DESKTOP: ~/$ cli <tab>
+# then: auto-complete fills in the very first match
+mauricel@DESKTOP: ~/$ cli book
+# when: user continues to press tab...
+mauricel@DESKTOP: ~/$ cli book<tab>
+# then: the user is presented with all options
+book        book add        book list       book remove     ...plus any files present
+mauricel@DESKTOP: ~/$ cli
+```
+
+# Architecture
+
+With CliFx, the auto-complete feature is comprised of the following components
+
+* terminal hook. This is typically an addition to a user's profile. The hook
+  * identifies the command to trigger upon 
+  * extracts cursor position and command line arguments 
+  * calls CliFx to determine the suggestions to provide suggestions
+* CliFx implements the `[suggest]` directive, which is called as follows:
+
+```PS
+# for CliFx.Demo book add title_of_book
+
+CliFx.Demo.exe [suggest] CliFx.Demo book add title_of_book
```

Comment:
Do we need to include `CliFx.Demo` after `[suggest]`? Given that it's already the application we're executing anyway.